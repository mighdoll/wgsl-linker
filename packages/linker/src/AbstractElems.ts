/** Structures for the abstract syntax tree constructed by the parser. */

import { FoundRef } from "./TraverseRefs.js";

export type AbstractElem =
  | AliasElem
  | ImportElem
  | ExtendsElem
  | ExportElem
  | ModuleElem
  | TemplateElem
  | FnElem
  | TypeNameElem
  | FnNameElem
  | VarNameElem
  | CallElem
  | StructElem
  | StructMemberElem
  | VarElem
  | TypeRefElem;

export type NamedElem = Extract<AbstractElem, { name: string }>;

export interface AbstractElemBase {
  kind: string;
  start: number;
  end: number;
}

export interface CallElem extends AbstractElemBase {
  kind: "call";
  name: string;
  ref?: FoundRef;
}

export interface FnNameElem extends AbstractElemBase {
  kind: "fnName";
  name: string;
}

export interface VarNameElem extends AbstractElemBase {
  kind: "varName";
  name: string;
}

export interface FnElem extends AbstractElemBase {
  kind: "fn";
  name: string;
  nameElem: FnNameElem;
  calls: CallElem[];
  typeRefs: TypeRefElem[];
}

export interface TypeRefElem extends AbstractElemBase {
  kind: "typeRef";
  name: string;
  ref?: FoundRef;
}

export interface TypeNameElem extends AbstractElemBase {
  kind: "typeName";
  name: string;
}

export interface StructElem extends AbstractElemBase {
  kind: "struct";
  name: string;
  nameElem: TypeNameElem;
  members: StructMemberElem[];
  typeRefs: TypeRefElem[];
  extendsElems?: ExtendsElem[];
}

export interface StructMemberElem extends AbstractElemBase {
  kind: "member";
  name: string;
}

export interface ExportElem extends AbstractElemBase {
  kind: "export";
  args?: string[];
  importing?: ImportElem[];
}

export interface ImportElem extends AbstractElemBase {
  kind: "import";
  name: string;
  args?: string[];
  as?: string;
  from?: string;
}

export interface ExtendsElem extends Omit<ImportElem, "kind"> {
  kind: "extends";
  name: string;
  args?: string[];
  as?: string;
  from?: string;
}

export interface ModuleElem extends AbstractElemBase {
  kind: "module";
  name: string;
}

export interface VarElem extends AbstractElemBase {
  kind: "var";
  name: string;
  nameElem: VarNameElem;
  typeRefs: TypeRefElem[];
}

export interface TemplateElem extends AbstractElemBase {
  kind: "template";
  name: string;
}

export interface AliasElem extends AbstractElemBase {
  kind: "alias";
  name: string;
  targetName: string;
  typeRefs: TypeRefElem[];
}