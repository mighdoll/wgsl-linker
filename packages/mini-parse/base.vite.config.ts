import { UserConfig } from "vite";
import tsconfigPaths from "vite-tsconfig-paths";
import dts from "vite-plugin-dts";
import { resolve } from "path";

export function baseViteConfig(): UserConfig {
  return {
    plugins: [
      tsconfigPaths(),
      dts(), // generate .d.ts files
    ],
    build: {
      lib: {
        name: "mini-parse",
        entry: [resolve(__dirname, "src/index.ts")],
      },
      minify: false,
      sourcemap: true,
    },
  };
}
