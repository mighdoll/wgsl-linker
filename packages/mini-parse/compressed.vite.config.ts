/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
// import { visualizer } from "rollup-plugin-visualizer";
import { baseViteConfig } from "./vite-base-config.js";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
