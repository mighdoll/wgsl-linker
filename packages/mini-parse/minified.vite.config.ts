/// <reference types="vitest" />
import { LibraryOptions, defineConfig } from "vite";
import { baseViteConfig } from "./base.vite.config.js";
// import { visualizer } from "rollup-plugin-visualizer";

// Note that this will include the debug tracing code, so
// so the size estimate is an overestimate of production size

const config = baseViteConfig();
config.build!.emptyOutDir = false;
config.build!.minify = "terser";
(config.build!.lib as LibraryOptions).formats = ["es", "cjs"];
(config.build!.lib as LibraryOptions).fileName = "minified";

// config.plugins?.push(visualizer({ brotliSize: true, gzipSize: true })); // generate stats.html size report

export default defineConfig(config);
