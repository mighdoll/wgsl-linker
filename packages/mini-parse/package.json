{
  "name": "mini-parse",
  "version": "0.2.0",
  "files": [
    "dist",
    "src"
  ],
  "scripts": {
    "pr": "printenv",
    "build": "run-s build:main build:util build:compressed",
    "build:main": "vite build",
    "build:util": "vite -c test-util.vite.config.ts build",
    "build:compressed": "vite -c compressed.vite.config.ts build",
    "build:brotli": "brotli dist/minified.cjs",
    "test": "vitest",
    "organize": "organize-imports-cli tsconfig.json",
    "format": "prettier-eslint \"src/**/*.ts\" --write",
    "lint": "eslint src"
  },
  "type": "module",
  "repository": "github:mighdoll/wgsl-linker",
  "exports": {
    ".": {
      "import": "./dist/index.js",
      "require": "./dist/index.cjs",
      "types": "./dist/index.d.ts"
    },
    "./minified": {
      "import": "./dist/minified.js",
      "require": "./dist/minified.cjs",
      "types": "./dist/index.d.ts"
    },
    "./test-util": {
      "import": "./dist/testUtil.js",
      "types": "./dist/test-util/testUtil.d.ts"
    }
  },
  "devDependencies": {
    "@types/node": "^20.11.16",
    "@typescript-eslint/eslint-plugin": "^6.21.0",
    "@typescript-eslint/parser": "^6.21.0",
    "@vitest/ui": "^1.1.1",
    "berry-pretty": "^0.0.4",
    "brotli": "^1.3.3",
    "esbuild": "^0.20.0",
    "eslint": "^8.56.0",
    "eslint-config-prettier": "^9.1.0",
    "npm-run-all": "^4.1.5",
    "organize-imports-cli": "^0.10.0",
    "prettier-eslint-cli": "^8.0.1",
    "rollup-plugin-visualizer": "^5.12.0",
    "terser": "^5.27.0",
    "typescript": "^5.3.3",
    "vite": "^5.0.12",
    "vite-plugin-dts": "^3.6.4",
    "vite-tsconfig-paths": "^4.2.1",
    "vitest": "^1.2.2"
  }
}
