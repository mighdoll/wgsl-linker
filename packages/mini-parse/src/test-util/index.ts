export * from "./LogCatcher.js";
export * from "./TestParse.js";
