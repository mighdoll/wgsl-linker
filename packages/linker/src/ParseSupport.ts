import {
  ExtendedResult,
  Parser,
  any,
  anyNot,
  disablePreParse,
  fn,
  kind,
  makeEolf,
  or,
  repeat,
  req,
  resultLog,
  seq,
  setTraceName,
  tracing,
  withSep,
} from "mini-parse";
import { AbstractElem } from "./AbstractElems.js";
import { argsTokens, mainTokens } from "./MatchWgslD.js";
import { lineCommentOptDirective } from "./ParseDirective.js";

/* Basic parsing functions for comment handling, eol, etc. */

export const word = kind(mainTokens.word);
export const wordNum = or(word, kind(mainTokens.digits));

export const unknown = any().map((r) => {
  const { kind, text } = r.value;
  resultLog(r, `??? ${kind}: '${text}'`);
});

export const blockComment: Parser<any> = seq(
  "/*",
  repeat(or(() => blockComment, anyNot("*/"))),
  req("*/")
);

export const comment = or(
  fn(() => lineCommentOptDirective),
  blockComment
);

export const eolf: Parser<any> = disablePreParse(
  makeEolf(argsTokens, argsTokens.ws)
);

/** ( a1, b1* ) with optinoal comments, spans lines */
export const wordNumArgs: Parser<string[]> = seq(
  "(",
  withSep(",", wordNum),
  req(")")
).map((r) => r.value[1]);

/** create an AbstractElem from parse results
 * @param named keys in the named result to copy to
 *  like named fields in the abstract elem (as a single value)
 * @param namedArray keys in the named result to copy to
 *  like named fields in the abstract elem (as an array)
 */
export function makeElem<U extends AbstractElem>(
  kind: U["kind"],
  er: ExtendedResult<any>,
  named: (keyof U)[] = [],
  namedArrays: (keyof U)[] = []
): U {
  const { start, end } = er;
  const nv = mapIfDefined(named, er.named as NameRecord<U>, true);
  const av = mapIfDefined(namedArrays, er.named as NameRecord<U>);
  return { kind, start, end, ...nv, ...av } as U;
}

type NameRecord<A> = Record<keyof A, string[]>;

function mapIfDefined<A>(
  keys: (keyof A)[],
  array: Record<keyof A, string[]>,
  firstElem?: boolean
): Partial<Record<keyof A, string>> {
  const entries = keys.flatMap((k) => {
    const ak = array[k];
    const v = firstElem ? ak?.[0] : ak;

    if (v === undefined) return [];
    else return [[k, v]];
  });
  return Object.fromEntries(entries);
}

// enableTracing();
if (tracing) {
  const names: Record<string, Parser<unknown>> = {
    skipBlockComment: blockComment,
    comment,
    wordNumArgs,
  };

  Object.entries(names).forEach(([name, parser]) => {
    setTraceName(parser, name);
  });
}
