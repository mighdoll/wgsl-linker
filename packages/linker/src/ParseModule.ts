import { srcLog, SrcMap } from "mini-parse";
import {
  AbstractElem,
  AliasElem,
  ExportElem,
  ExtendsElem,
  FnElem,
  GlobalDirectiveElem,
  ImportElem,
  ModuleElem,
  StructElem,
  TemplateElem,
  VarElem,
} from "./AbstractElems.js";
import { processConditionals } from "./Conditionals.js";
import { ApplyTemplateFn } from "./ModuleRegistry.js";
import { parseWgslD } from "./ParseWgslD.js";
import { SliceReplace, sliceReplace } from "./Slicer.js";

/** module with exportable text fragments that are optionally transformed by a templating engine */
export interface TextModule {
  kind: "text";
  template?: TemplateElem;
  exports: TextExport[];
  fns: FnElem[];
  vars: VarElem[];
  structs: StructElem[];
  imports: (ImportElem | ExtendsElem)[];
  aliases: AliasElem[];
  globalDirectives: GlobalDirectiveElem[];

  /** name of the module. A synthetic name will be assigned if none is provided */
  name: string;

  /** name of the module. A synthetic file name will be assigned if none is provided */
  fileName: string;

  /** original src for module */
  src: string;

  /** src code after processing #if conditionals  */
  preppedSrc: string;

  /** tracks changes through conditional processing for error reporting */
  srcMap: SrcMap;
}

/** an export elem annotated with the fn/struct to which it refers */
export interface TextExport extends ExportElem {
  ref: FnElem | StructElem;
}

let unnamedModuleDex = 0;
let unnamedFileDex = 0;

export function preProcess(
  src: string,
  params: Record<string, any> = {},
  templates: Map<string, ApplyTemplateFn> = new Map()
): SrcMap {
  const condSrcMap = processConditionals(src, params);
  return applyTemplate(condSrcMap, templates, params);
}

export function parseModule(
  src: string,
  templates: Map<string, ApplyTemplateFn> = new Map(),
  fileName = `/unnamed-${unnamedFileDex++}`,
  params: Record<string, any> = {},
  defaultModuleName?: string
): TextModule {
  const srcMap = preProcess(src, params, templates);

  const preppedSrc = srcMap.dest;
  const parsed = parseWgslD(preppedSrc, srcMap);
  const exports = findExports(parsed, srcMap);
  const fns = filterElems<FnElem>(parsed, "fn");
  const aliases = filterElems<AliasElem>(parsed, "alias");
  const globalDirectives = filterElems<GlobalDirectiveElem>(
    parsed,
    "globalDirective"
  );
  const imports = parsed.filter(
    (e) => e.kind === "import" || e.kind === "extends"
  ) as (ImportElem | ExtendsElem)[];
  const structs = filterElems<StructElem>(parsed, "struct");
  const vars = filterElems<VarElem>(parsed, "var");
  const template = filterElems<TemplateElem>(parsed, "template")?.[0];
  matchMergeImports(parsed, srcMap);
  const moduleName = filterElems<ModuleElem>(parsed, "module")[0]?.name;
  matchMergeImports(parsed, srcMap);

  const name = moduleName ?? defaultModuleName ?? `module${unnamedModuleDex++}`;
  const kind = "text";
  return {
    ...{ kind, src, srcMap, preppedSrc, fileName, name },
    ...{
      exports,
      fns,
      structs,
      vars,
      imports,
      template,
      aliases,
      globalDirectives,
    },
  };
}

export function filterElems<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"]
): T[] {
  return parsed.filter((e) => e.kind === kind) as T[];
}

function findExports(parsed: AbstractElem[], srcMap: SrcMap): TextExport[] {
  const results: TextExport[] = [];
  const exports = findKind<ExportElem>(parsed, "export");

  exports.forEach(([elem, i]) => {
    let next: AbstractElem | undefined;
    do {
      next = parsed[++i];
    } while (next?.kind === "extends");
    if (elem.kind === "export") {
      if (next?.kind === "fn" || next?.kind === "struct") {
        results.push({ ...elem, ref: next });
      } else {
        srcLog(srcMap, elem.start, `#export what? (#export a fn or struct)`);
      }
    }
  });
  return results;
}

/** fill in extendsElem field of structs */
function matchMergeImports(parsed: AbstractElem[], srcMap: SrcMap): void {
  const extendsElems = findKind<ExtendsElem>(parsed, "extends");
  extendsElems.forEach(([extendsElem, i]) => {
    let next: AbstractElem | undefined;
    do {
      next = parsed[++i];
    } while (next?.kind === "extends" || next?.kind === "export");
    if (next?.kind === "struct") {
      next.extendsElems = next.extendsElems ?? [];
      next.extendsElems.push(extendsElem);
    } else {
      srcLog(srcMap, extendsElem.start, `#extends not followed by a struct`);
    }
  });
}

function findKind<T extends AbstractElem>(
  parsed: AbstractElem[],
  kind: T["kind"]
): [T, number][] {
  return parsed.flatMap((elem, i) =>
    elem.kind === kind ? ([[elem, i]] as [T, number][]) : []
  );
}

const templateRegex = /#template\s+([/[a-zA-Z_][\w./-]*)/;

function applyTemplate(
  priorSrcMap: SrcMap,
  templates: Map<string, ApplyTemplateFn>,
  params: Record<string, any>
): SrcMap {
  const src = priorSrcMap.dest;
  const foundTemplate = src.match(templateRegex);
  if (!foundTemplate) {
    return priorSrcMap;
  }
  const templateName = foundTemplate[1];
  const templateFn = templates.get(templateName);
  if (!templateFn) {
    srcLog(
      priorSrcMap,
      foundTemplate.index!,
      `template '${templateName}' not found in ModuleRegistry`
    );
    return priorSrcMap;
  }

  // dlog({priorSrcMap})

  const start = foundTemplate.index!;
  const end = start + foundTemplate[0].length;
  const rmDirective: SliceReplace = { start, end, replacement: "" };
  const removedMap = sliceReplace(src, [rmDirective]);
  // dlog({ removedMap });
  const removeMerged = priorSrcMap.merge(removedMap);
  // dlog({ removeMerged });

  const templatedMap = templateFn(removeMerged.dest, params);
  // dlog({ templatedMap});

  const srcMap = removeMerged.merge(templatedMap);
  // dlog({ srcMap });

  return srcMap;
}
