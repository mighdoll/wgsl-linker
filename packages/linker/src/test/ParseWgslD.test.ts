import { _withBaseLogger, or, repeat } from "mini-parse";
import { expectNoLogErr, logCatch } from "mini-parse/test-util";

import { expect, test } from "vitest";
import { AbstractElem, FnElem, StructElem, VarElem } from "../AbstractElems.js";
import { filterElems } from "../ParseModule.js";
import { unknown, wordNumArgs } from "../ParseSupport.js";
import {
  fnDecl,
  globalVar,
  parseWgslD,
  structDecl,
  typeSpecifier,
} from "../ParseWgslD.js";
import { testAppParse } from "./TestUtil.js";

function testParseWgsl(src: string): AbstractElem[] {
  return parseWgslD(src, undefined, {}, 500);
}

test("parse empty string", () => {
  const parsed = testParseWgsl("");
  expect(parsed).toMatchSnapshot();
});

test("parse fn foo() { }", () => {
  const src = "fn foo() { }";
  const parsed = testParseWgsl(src);
  expect(parsed).toMatchSnapshot();
});

test("parse fn with calls", () => {
  const src = "fn foo() { foo(); bar(); }";
  const parsed = testParseWgsl(src);
  expect(parsed).toMatchSnapshot();
});

test("structDecl parses struct member types", () => {
  const src = "struct Foo { a: f32, b: i32 }";
  const { appState } = testAppParse(structDecl, src);
  const { members } = appState[0] as StructElem;
  const typeNames = members.flatMap((m) => m.typeRefs.map((t) => t.name));
  expect(typeNames).deep.eq(["f32", "i32"]);
});

test("parse struct", () => {
  const src = "struct Foo { a: f32, b: i32 }";
  const parsed = testParseWgsl(src);
  expect(parsed).toMatchInlineSnapshot(`
    [
      {
        "end": 29,
        "kind": "struct",
        "members": [
          {
            "end": 19,
            "kind": "member",
            "name": "a",
            "start": 13,
            "typeRefs": [
              {
                "end": 19,
                "kind": "typeRef",
                "name": "f32",
                "start": 16,
              },
            ],
          },
          {
            "end": 27,
            "kind": "member",
            "name": "b",
            "start": 21,
            "typeRefs": [
              {
                "end": 27,
                "kind": "typeRef",
                "name": "i32",
                "start": 24,
              },
            ],
          },
        ],
        "name": "Foo",
        "nameElem": {
          "end": 10,
          "kind": "typeName",
          "name": "Foo",
          "start": 7,
        },
        "start": 0,
      },
    ]
  `);
});

test("parse @attribute before fn", () => {
  const src = `
    @compute 
    fn main() {}
    `;
  const parsed = testParseWgsl(src);
  expect(parsed).toMatchSnapshot();
});

test("wordNumArgs parses (a, b, 1)", () => {
  const src = `(a, b, 1)`;
  const { parsed } = testAppParse(wordNumArgs, src);
  expect(parsed?.value).toMatchSnapshot();
});

test("parse @compute @workgroup_size(a, b, 1) before fn", () => {
  const src = `
    @compute 
    @workgroup_size(a, b, 1) 
    fn main() {}
    `;
  const parsed = testParseWgsl(src);
  expect(parsed).toMatchSnapshot();
});

test("parse global diagnostic", () => {
  const src = `
    diagnostic(off,derivative_uniformity);

    fn main() {}
    `;
  expectNoLogErr(() => {
    const parsed = testParseWgsl(src);
    expect(parsed).toMatchSnapshot();
  });
});

test("parse const_assert", () => {
  const src = `
    const_assert x < y;

    fn main() {}
    `;
  expectNoLogErr(() => {
    const parsed = testParseWgsl(src);
    expect(parsed).toMatchSnapshot();
  });
});

test("parse top level var", () => {
  const src = `
    @group(0) @binding(0) var<uniform> u: Uniforms;      

    fn main() {}
  `;
  expectNoLogErr(() => {
    const parsed = testParseWgsl(src);
    expect(parsed).toMatchSnapshot();
  });
});

test("parse top level override and const", () => {
  const src = `
    override x = 21;
    const y = 1;

    fn main() {}
  `;
  expectNoLogErr(() => {
    const parsed = testParseWgsl(src);
    expect(parsed).toMatchSnapshot();
  });
});

test("parse root level ;;", () => {
  const src = ";;";
  expectNoLogErr(() => {
    const parsed = testParseWgsl(src);
    expect(parsed).toMatchSnapshot();
  });
});

test("parse simple alias", () => {
  const src = `alias NewType = OldType;`;
  expectNoLogErr(() => {
    const parsed = testParseWgsl(src);
    expect(parsed).toMatchSnapshot();
  });
});

test("parse array alias", () => {
  const src = `
    alias Points3 = array<Point, 3>;
  `;
  expectNoLogErr(() => {
    const parsed = testParseWgsl(src);
    expect(parsed).toMatchSnapshot();
  });
});

test("unexpected token", () => {
  const p = repeat(or("a", unknown));
  const { log, logged } = logCatch();
  _withBaseLogger(log, () => testAppParse(p, "a b"));
  expect(logged()).toMatchInlineSnapshot(`
    "??? word: 'b'
    a b   Ln 1
      ^"
  `);
});

test("fnDecl parses fn with return type", () => {
  const src = `
    fn foo() -> MyType { }
  `;
  const { appState } = testAppParse(fnDecl, src);
  expect((appState[0] as FnElem).typeRefs[0].name).eq("MyType");
});

test("fnDecl parses :type specifier in fn args", () => {
  const src = `
    fn foo(a: MyType) { }
  `;
  const { appState } = testAppParse(fnDecl, src);
  const { typeRefs } = appState[0] as FnElem;
  expect(typeRefs[0].name).eq("MyType");
});

test("fnDecl parses :type specifier in fn block", () => {
  const src = `
    fn foo() { 
      var b:MyType;
    }
  `;
  const { appState } = testAppParse(fnDecl, src);
  expect((appState[0] as FnElem).typeRefs[0].name).eq("MyType");
});

test("parse type in <template> in fn args", () => {
  const src = `
    fn foo(a: vec2<MyStruct>) { };`;

  const { appState } = testAppParse(fnDecl, src);
  const { typeRefs } = appState[0] as FnElem;
  expect(typeRefs[0].name).eq("vec2");
  expect(typeRefs[1].name).eq("MyStruct");
});

test("parse simple templated type", () => {
  const src = `array<MyStruct,4>`;

  const { parsed } = testAppParse(typeSpecifier, src);
  expect(parsed?.value[0].name).eq("array");
  expect(parsed?.value[1].name).eq("MyStruct");
  expect(parsed?.value.length).eq(2);
});

test("parse nested template that ends with >> ", () => {
  const src = `vec2<array <MyStruct,4>>`;

  const { parsed } = testAppParse(typeSpecifier, src);
  const typeRefNames = parsed?.value.map((r) => r.name);
  expect(typeRefNames).deep.eq(["vec2", "array", "MyStruct"]);
});

test("parse struct member with templated type", () => {
  const src = `struct Foo { a: vec2<array<Bar,4>> }`;
  const { appState } = testAppParse(structDecl, src);
  const members = filterElems<StructElem>(appState, "struct")[0].members;
  const memberNames = members.flatMap((m) => m.typeRefs.map((t) => t.name));
  expect(memberNames).deep.eq(["vec2", "array", "Bar"]);
});

test("parse type in <template> in global var", () => {
  const src = `
    var x:vec2<MyStruct> = { x: 1, y: 2 };`;

  const { appState } = testAppParse(globalVar, src);
  const typeRefs = (appState[0] as VarElem).typeRefs;
  expect(typeRefs[0].name).eq("vec2");
  expect(typeRefs[1].name).eq("MyStruct");
});

test("parse for(;;) {} not as a fn call", () => {
  const src = `
    fn main() {
      for (var a = 1; a < 10; a++) {}
    }
  `;
  const appState = testParseWgsl(src);
  const fnElem = filterElems<FnElem>(appState, "fn")[0];
  expect(fnElem).toBeDefined();
  expect(fnElem.calls.length).eq(0);
});

test("eolf followed by blank line", () => {
  const src = `
    // #if typecheck
    // #endif

    // #export
    fn foo() { }
  `;
  expectNoLogErr(() => testParseWgsl(src));
});

test("parse fn with attributes and suffix comma", () => {
  const src = `
  @compute
  @workgroup_size(workgroupThreads, 1, 1) 
  fn main(
      @builtin(global_invocation_id) grid: vec3<u32>,
      @builtin(local_invocation_index) localIndex: u32,  
  ) { }
  `;
  expectNoLogErr(() => {
    const parsed = testParseWgsl(src);
    const first = parsed[0] as FnElem;
    expect(first.kind).eq("fn");
    expect(first.name).eq("main");
  });
});
