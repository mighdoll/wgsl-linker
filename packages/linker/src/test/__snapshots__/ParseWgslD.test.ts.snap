// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse @attribute before fn 1`] = `
[
  {
    "calls": [],
    "end": 31,
    "kind": "fn",
    "name": "main",
    "nameElem": {
      "end": 26,
      "kind": "fnName",
      "name": "main",
      "start": 22,
    },
    "start": 5,
    "typeRefs": [],
  },
]
`;

exports[`parse @compute @workgroup_size(a, b, 1) before fn 1`] = `
[
  {
    "calls": [],
    "end": 61,
    "kind": "fn",
    "name": "main",
    "nameElem": {
      "end": 56,
      "kind": "fnName",
      "name": "main",
      "start": 52,
    },
    "start": 5,
    "typeRefs": [],
  },
]
`;

exports[`parse array alias 1`] = `
[
  {
    "end": 37,
    "kind": "alias",
    "name": "Points3",
    "start": 5,
    "typeRefs": [
      {
        "end": 36,
        "kind": "typeRef",
        "name": "array",
        "start": 21,
      },
      {
        "end": 36,
        "kind": "typeRef",
        "name": "Point",
        "start": 21,
      },
    ],
  },
]
`;

exports[`parse const_assert 1`] = `
[
  {
    "end": 24,
    "kind": "globalDirective",
    "start": 5,
  },
  {
    "calls": [],
    "end": 42,
    "kind": "fn",
    "name": "main",
    "nameElem": {
      "end": 37,
      "kind": "fnName",
      "name": "main",
      "start": 33,
    },
    "start": 30,
    "typeRefs": [],
  },
]
`;

exports[`parse empty string 1`] = `[]`;

exports[`parse fn foo() { } 1`] = `
[
  {
    "calls": [],
    "end": 12,
    "kind": "fn",
    "name": "foo",
    "nameElem": {
      "end": 6,
      "kind": "fnName",
      "name": "foo",
      "start": 3,
    },
    "start": 0,
    "typeRefs": [],
  },
]
`;

exports[`parse fn with calls 1`] = `
[
  {
    "calls": [
      {
        "end": 14,
        "kind": "call",
        "name": "foo",
        "start": 11,
      },
      {
        "end": 21,
        "kind": "call",
        "name": "bar",
        "start": 18,
      },
    ],
    "end": 26,
    "kind": "fn",
    "name": "foo",
    "nameElem": {
      "end": 6,
      "kind": "fnName",
      "name": "foo",
      "start": 3,
    },
    "start": 0,
    "typeRefs": [],
  },
]
`;

exports[`parse foo.bar();  1`] = `
[
  {
    "calls": [
      {
        "end": 19,
        "kind": "call",
        "name": "foo.bar",
        "start": 12,
      },
    ],
    "end": 24,
    "kind": "fn",
    "name": "main",
    "nameElem": {
      "end": 7,
      "kind": "fnName",
      "name": "main",
      "start": 3,
    },
    "start": 0,
    "typeRefs": [],
  },
]
`;

exports[`parse foo::bar();  1`] = `
[
  {
    "calls": [
      {
        "end": 20,
        "kind": "call",
        "name": "foo::bar",
        "start": 12,
      },
    ],
    "end": 25,
    "kind": "fn",
    "name": "main",
    "nameElem": {
      "end": 7,
      "kind": "fnName",
      "name": "main",
      "start": 3,
    },
    "start": 0,
    "typeRefs": [],
  },
]
`;

exports[`parse global diagnostic 1`] = `
[
  {
    "end": 43,
    "kind": "globalDirective",
    "start": 5,
  },
  {
    "calls": [],
    "end": 61,
    "kind": "fn",
    "name": "main",
    "nameElem": {
      "end": 56,
      "kind": "fnName",
      "name": "main",
      "start": 52,
    },
    "start": 49,
    "typeRefs": [],
  },
]
`;

exports[`parse root level ;; 1`] = `[]`;

exports[`parse simple alias 1`] = `
[
  {
    "end": 24,
    "kind": "alias",
    "name": "NewType",
    "start": 0,
    "typeRefs": [
      {
        "end": 23,
        "kind": "typeRef",
        "name": "OldType",
        "start": 16,
      },
    ],
  },
]
`;

exports[`parse top level override and const 1`] = `
[
  {
    "end": 21,
    "kind": "var",
    "name": "x",
    "start": 5,
    "typeRefs": [],
  },
  {
    "end": 38,
    "kind": "var",
    "name": "y",
    "start": 26,
    "typeRefs": [],
  },
  {
    "calls": [],
    "end": 56,
    "kind": "fn",
    "name": "main",
    "nameElem": {
      "end": 51,
      "kind": "fnName",
      "name": "main",
      "start": 47,
    },
    "start": 44,
    "typeRefs": [],
  },
]
`;

exports[`parse top level var 1`] = `
[
  {
    "end": 52,
    "kind": "var",
    "name": "u",
    "start": 5,
    "typeRefs": [
      {
        "end": 51,
        "kind": "typeRef",
        "name": "Uniforms",
        "start": 43,
      },
    ],
  },
  {
    "calls": [],
    "end": 76,
    "kind": "fn",
    "name": "main",
    "nameElem": {
      "end": 71,
      "kind": "fnName",
      "name": "main",
      "start": 67,
    },
    "start": 64,
    "typeRefs": [],
  },
]
`;

exports[`wordNumArgs parses (a, b, 1) 1`] = `
[
  "a",
  "b",
  "1",
]
`;
